File: backend/app/services/user_service.py
"""ArthaNova — User service layer."""

from typing import Optional
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select
from datetime import datetime, timezone

from app.models.user import User, Portfolio
from app.core.security import get_password_hash, verify_password
from app.schemas.schemas import UserRegisterRequest


async def get_user_by_id(db: AsyncSession, user_id: int) -> Optional[User]:
    result = await db.execute(select(User).where(User.id == user_id))
    return result.scalar_one_or_none()


async def get_user_by_email(db: AsyncSession, email: str) -> Optional[User]:
    result = await db.execute(select(User).where(User.email == email))
    return result.scalar_one_or_none()


async def get_user_by_username(db: AsyncSession, username: str) -> Optional[User]:
    result = await db.execute(select(User).where(User.username == username))
    return result.scalar_one_or_none()


async def create_user(db: AsyncSession, data: UserRegisterRequest) -> User:
    user = User(
        email=data.email,
        username=data.username,
        full_name=data.full_name,
        phone=data.phone,
        hashed_password=get_password_hash(data.password),
        is_active=True,
        is_verified=False,
    )
    db.add(user)
    await db.flush()

    # Create default portfolio
    portfolio = Portfolio(user_id=user.id, name=f"{data.full_name}'s Portfolio")
    db.add(portfolio)
    await db.flush()

    return user


async def authenticate_user(db: AsyncSession, email: str, password: str) -> Optional[User]:
    user = await get_user_by_email(db, email)
    if not user:
        return None
    if not verify_password(password, user.hashed_password):
        return None
    return user


async def update_last_login(db: AsyncSession, user: User):
    user.last_login = datetime.now(timezone.utc)
    await db.flush()


async def update_user_password(db: AsyncSession, user: User, new_password: str):
    user.hashed_password = get_password_hash(new_password)
    await db.flush()

File: backend/src/services/userService.js
import db from '../models/db.js';

export const getUserByEmail = async (email) => {
  return await db.queryFirst('SELECT * FROM users WHERE email = ?', [email]);
};

export const getUserByUsername = async (username) => {
  return await db.queryFirst('SELECT * FROM users WHERE username = ?', [username]);
};

export const getUserById = async (id) => {
  return await db.queryFirst('SELECT * FROM users WHERE id = ?', [id]);
};

export const createUser = async (userData) => {
  const { email, username, full_name, hashed_password } = userData;
  const result = await db.execute(
    'INSERT INTO users (email, username, full_name, hashed_password) VALUES (?, ?, ?, ?)',
    [email, username, full_name, hashed_password]
  );
  // Fetch and return the newly created user
  return await db.queryFirst('SELECT * FROM users WHERE id = ?', [Number(result.lastInsertRowid)]);
};

export const updateUserLastLogin = async (id) => {
  const now = new Date().toISOString();
  return await db.execute(
    'UPDATE users SET last_login = ?, updated_at = ? WHERE id = ?',
    [now, now, id]
  );
};

export const updateUserPassword = async (id, hashedPassword) => {
  const now = new Date().toISOString();
  return await db.execute(
    'UPDATE users SET hashed_password = ?, updated_at = ? WHERE id = ?',
    [hashedPassword, now, id]
  );
};
export const updateUserAdminStatus = async (id, isAdmin) => {
  const now = new Date().toISOString();
  return await db.execute(
    'UPDATE users SET is_admin = ?, updated_at = ? WHERE id = ?',
    [isAdmin ? 1 : 0, now, id]
  );
};
export const logAction = async (data) => {
  const { user_id, email, action, module = 'CORE', details, status = 'SUCCESS', ip } = data;
  try {
    const timestamp = new Date().toISOString();
    return await db.execute(
      `INSERT INTO audit_logs (user_id, email, action, module, details, status, ip_address, timestamp) 
       VALUES (?, ?, ?, ?, ?, ?, ?, ?)`,
      [user_id, email, action, module, details, status, ip, timestamp]
    );
  } catch (err) {
    console.error('FAILED TO LOG AUDIT:', err);
  }
};
